### PR TITLE
RFC: Remove 'auto refresh' feature

### PR DESCRIPTION
**Pull request by daniel-beck**
**Request for comments**: How do we feel about getting rid of "auto refresh"?

I've hated this feature for a long time:

1. It's a terrible implementation of keeping the view updated; forcing the entire page to reload every N seconds while enabled, keeping state such as scroll position only if lucky and burdening the Jenkins server, if an expensive enough view is shown.
2. It needs special support in widgets that otherwise allow inline editing, just because they may be used on pages that are auto-reloadable.
3. It can cause data loss if a view containing input does not disable auto refresh.

Considering all the above, I think this feature is a net negative.

Users who want to get the complete view updated are probably better off using any of the radiator/wall display plugin, or Blue Ocean, who are designed for this use case.

### Proposed changelog entries

* Remove "auto refresh" functionality.

### Desired reviewers

@jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/resources/lib/layout/html.jelly [hunk 1]
```diff
@@ -37,8 +37,7 @@ THE SOFTWARE.
       Title of the HTML page. Rendered into &lt;title> tag.
     </st:attribute>
     <st:attribute name="norefresh">
```
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/html.jelly]: There's an actual `deprecated` attribute for the tag.

**Comment by oleg-nenashev** [comment]
Weak +1 for removing the feature. I also see no real use-case for it now.
My proposal would be to mark the feature as deprecated and to move it to the feature flag (e.g. in Global config or in System property), which would be handled in `isAutoRefresh()` then

**Comment by daniel-beck** [comment]
@oleg-nenashev This is a feature that depends on ever plugin developer being aware of it to support it in their plugins, so that seems like the most unfeasible way forward.

**Comment by rtyler** [comment]
:+1:  kill it with fire, terribad parts of Jenkins should be removed post-haste

raw

**Comment by MarkEWaite** [comment]
I'm fine with it being removed.

**Comment by recena** [comment]
👍

File: core/src/main/resources/lib/layout/html.jelly (hunk 1 above)
Comment (recena) [inline core/src/main/resources/lib/layout/html.jelly]: @daniel-beck As far as I remember, this Jelly tag is deprecated and suggested for removing (https://github.com/jenkinsci/jenkins/pull/2852)

**Review by oleg-nenashev — APPROVED** [review]
It got enough reviewers. I would rather vote for a final sign-off at the Aug 17 governance meeting, but the people always say I misuse these meetings

**Comment by rtyler** [comment]
@oleg-nenashev heh, I think you're just on the more conservative side of the developer-spectrum here :smile: 

If removing any minor feature required an agenda in the governance meeting, we'll never remove all these bad old features and bugs.

I think it would be worth @daniel-beck  sending a heads up to the dev list with a time delay "if nobody cares in 3 days, I'll merge"

File: core/src/main/resources/lib/layout/breadcrumbBar.jelly [hunk 2]
```diff
@@ -40,20 +40,7 @@ THE SOFTWARE.
...
-                      <a href="?auto_refresh=false">${%DISABLE AUTO REFRESH}</a>
-                    </j:when>
-                    <j:otherwise>
-                      <a href="?auto_refresh=true">${%ENABLE AUTO REFRESH}</a>
-                    </j:otherwise>
-                  </j:choose>
-                </div>
-              </div>
-            </j:if>
+          <!-- used to contain auto refresh options, kept for layout backward compatibility -->
```
Comment (recena) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @daniel-beck Who is using this element `<div id="right-top-nav">`?

Comment (daniel-beck) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @recena I don't know and don't care enough to find out.

Comment (recena) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @daniel-beck then I'd remove the node `<div id="right-top-nav">`

Comment (daniel-beck) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: To clarify, without knowing with reasonable certainty I prefer to keep an unused container over breaking plugins. I just don't care enough about cleaning up the HTML to spend the time finding out.

Comment (recena) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @daniel-beck IMHO, this code should be cleaned up. If a plugin is assuming this HTML node is present, it is a very bad practice.

Comment (recena) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @daniel-beck What do you mean with _I just don't care enough about cleaning up the HTML to spend the time finding out_? Is it not important for you to keep the UI code organized and clean?

Comment (daniel-beck) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: Not particularly. It's important to me to keep it functional and prevent regressions.

This is a separate discussion from what this PR is trying to accomplish, could we keep this around and do this in a separately evaluated followup change?

Comment (recena) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @daniel-beck I don't know why this should be managed in a different PR. It is part of removing a feature / functionality.

Anyway, I'm not blocking anything, basically, I'm telling what I would do.

**Comment by oleg-nenashev** [comment]
@daniel-beck so do you plan to send a message to the Dev ML as @rtyler proposed? It sound to be a good approach to get it in the next weekly (if we really feel brave to land something big before JW)

**Comment by daniel-beck** [comment]
Invited feedback at https://groups.google.com/d/msg/jenkinsci-dev/rKV3omz1X60/lfmci6aICAAJ

**Comment by slide** [comment]
:+1:

**Comment by aheritier** [comment]
👍

**Review by aheritier — APPROVED** [review]

**Comment by rpionke** [comment]
:+1:

**Review by olivergondza — APPROVED** [review]

**Comment by i386** [comment]
+1 to killing it.

**Comment by orrc** [comment]
I've never found it useful, so I'm in favour of killing it — but I'm sure a vocal minority of people will miss it, but in that case we should suggest they try the various browser plugins that implement this functionality.

**Comment by daniel-beck** [comment]
> the various browser plugins that implement this functionality.

Good point. This is literally only a browser reload every ten seconds (disabled on some pages). It'll even be configurable interval per-user 👍

File: core/src/main/resources/lib/layout/breadcrumbBar.jelly (hunk 2 above)
Comment (jglick) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: I can find no usages so would be fine removing it, but it seems harmless enough to leave it there.

File: core/src/main/resources/lib/layout/html.jelly [hunk 3]
```diff
@@ -37,8 +37,7 @@ THE SOFTWARE.
...
-      If non-null and not "false", auto refresh is disabled on this page.
-      This is necessary for pages that include forms.
+      Deprecated: Used to disable auto refresh for a page. The feature has been removed.
```
Comment (jglick) [inline core/src/main/resources/lib/layout/html.jelly]: Are you deliberately not removing `norefresh` from the many places it is specified, such as `CLIAction/command.jelly` for example?

File: core/src/main/resources/lib/layout/layout.jelly [hunk 4]
```diff
@@ -79,7 +79,6 @@ ${h.initPageVariables(context)}
```
Comment (jglick) [inline core/src/main/resources/lib/layout/layout.jelly]: You left `norefresh` undeprecated in this file! The one which is actually used, unlike `html.jelly`.

**Review by jglick — COMMENTED** [review]
:+1: generally but you seemed to have missed a few things. Recommend grepping the whole repo for `efresh` to see.

File: core/src/main/resources/lib/layout/breadcrumbBar.jelly (hunk 2 above)
Comment (recena) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: I don't know why many contributors of this project look like _obsessed_ with keeping code that is not used 😄

Comment (MarkEWaite) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @recena I'm _obsessed_ with keeping code that is not used because of all the times I've been proven wrong when I thought code was not being used.  My obsession will likely continue 😄

Comment (recena) [inline core/src/main/resources/lib/layout/breadcrumbBar.jelly]: @MarkEWaite Remove code is part of the software engineering 😄

**Comment by oleg-nenashev** [comment]
Some notes: 

* What will happen with the startup page ("Your browser will reload automatically when Jenkins is ready."). Does it depend on the auto-refresh?
* Browser plugin is probably not enough if it does not support auto-refreshing **only** for particular pages

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> Does it depend on the auto-refresh?

No.

**Review by jetersen — APPROVED** [review]
KILL WITH 🚀

**Comment by oleg-nenashev** [comment]
Ready to go

**Comment by oleg-nenashev** [comment]
@daniel-beck You still prefer to keep it unmerged, right?

**Comment by daniel-beck** [comment]
@oleg-nenashev Yes, didn't get to it this week.

File: core/src/main/resources/lib/layout/layout.jelly [hunk 5]
```diff
@@ -34,8 +34,7 @@ THE SOFTWARE.
       Title of the HTML page. Rendered into &lt;title> tag.
     </st:attribute>
     <st:attribute name="norefresh">
-      If non-null and not "false", auto refresh is disabled on this page.
-      This is necessary for pages that include forms.
+      Deprecated: Used to disable auto refresh for a page. The feature has been removed.
```
Comment (recena) [inline core/src/main/resources/lib/layout/layout.jelly]: @daniel-beck What do you think if we add something like `@since TODO`?

**Review by batmat — APPROVED** [review]

**Comment by batmat** [comment]
@daniel-beck conflicted.

**Comment by daniel-beck** [comment]
@batmat we have a separate label for that.

**Review by Wadeck — APPROVED** [review]

**Review by rpionke — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Should it be JEP now? Not mandatory since there is a wide approval in pre-JEP discussions tho

**Comment by daniel-beck** [comment]
@oleg-nenashev Good question. I'm not sure. Will need to consider approach when I have some time.

**Comment by slide** [comment]
I don't think this really needs to be a JEP, the feature isn't THAT big that is being removed. This seems too small to necessitate a JEP.

**Comment by oleg-nenashev** [comment]
@daniel-beck are you still interested in it?

**Comment by daniel-beck** [comment]
@oleg-nenashev Interested definitely. If there's consensus it's mergeable after just some conflict resolution I can do that next week.

**Comment by oleg-nenashev** [comment]
@daniel-beck yes, I believe we have a consensus here, everybody voted +1. And I believe there is also a consensus that JEP is not needed

@jenkinsci/core Please respond to this PR if you disagree with removing autorefresh

**Comment by batmat** [comment]
Agreed that this should be merged.

**Comment by daniel-beck** [comment]
Will re-do after ##3894.
